---
title: "Identifying SARS-CoV-2 mutations in Beijing and transmission rates in China in 2020"
author: "Sarah Gao"
date: "November 19, 2020"
output:
  pdf_document: default
  html_document:
    df_print: paged
csl: bioinformatics.csl
bibliography: references.bib
params:
  gff_file_path: /data/sars_vcf_analysis/02_genome_reference/sars_refgenome_annotation.gff
  vcf_dir_path: data/11_vcf_output_for_R
  sra_runtable_path: data/00_sra_runtable/SraRunTable_PRJNA667180.txt
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      warning = FALSE,
                      message = FALSE,
                      error = TRUE)
```

# Background and Overview

In late 2019, the first cases of Severe Acute Respiratory Syndrome Coronavirus 2 caused by the SARS-CoV-2 virus were reported in Wuhan, China. Since then, the virus has spread rapidly around the world and has resulted in more than 1.4 million deaths worldwide as of November, 27th 2020 [@jhu2020].

Researchers around the globe have identified key mutations that are predominant in different locations, indicating that the virus has been evolving over time to form location-specific strains. This report examined how the SARS-CoV-2 virus has mutated within and affected the population in Beijing, China. Beijing is approximately 717 miles (~1,153 km) from Wuhan, a city in Hubei province that was the center of the first known COVID-19 cases first reported in December 2019.

The goal of this report was to identify mutations of the SARS-CoV-2 virus found in samples taken from positive patients in a Beijing hospital, compare them with known mutations identified by previous researchers, and analyze data on China's response to the virus at a national level. I used a pipeline written in bash to process the sequence data and identify variants from the reference Wuhan genome. I called out specifically variants at positions that are commonly found in samples around the world in order to examine how Beijing's strain may have mutated since the virus first emerged [@pachetti2020][@vanDorp2020].

Previous studies showed that a mutation in the gene encoding the S spike protein has created a SARS-CoV-2 variant that has become the most prevalent form in many geographies, suggesting that this strain could have an evolutionary advantage [@korber2020]. While there has been no definitive understanding in how known mutations in the virus affect the virus' transmission or infection capacity, it is certainly an important area of study to understand how the virus will change over time, and in turn, how we may more effectively combat it. Analyses of this Beijing dataset show that there were no more S gene mutations than one would expect assuming equal rates of mutation across the genome. However, there are many other factors that can affect how a region copes with the ever-changing COVID-19 pandemic other than the potential outcomes of viral mutations, including population and government responses. I found that the number of cases across the country as a whole has reduced dramatically since spring of 2020, despite fluctuating effective reproduction rates, as the Chinese government has reacted quickly and strictly in response to spikes. 
  
\newpage  
  
# Methods

## SARS-CoV-2 Sequence Data
I downloaded the Beijing dataset from the [NCBI Database](https://www.ncbi.nlm.nih.gov/Traces/study/?acc=PRJNA667180&o=acc_s%3Aa) on Wednesday November 18th, 2020. This dataset was found originally by looking through [NCBI's SRA BioProjects](https://www.ncbi.nlm.nih.gov/bioproject/?term=(txid2697049[Organism:noexp]%20NOT%200[Mbases])%20AND%20bioproject_sra[filter]%20NOT%20bioproject_gap[filter]) and filtering for Illumina sequences only. The 102 samples in the dataset were taken from SARS-CoV-2 positive patients at Beijing Ditan Hospital Capital Medical University between January 29th and April 17th, 2020 via three isolation sources: feces, pharyngeal swab, and sputum (Figure 1) [@du2020].

I then downloaded the SARS-CoV-2 Wuhan reference genome from [NCBI](https://www.ncbi.nlm.nih.gov/sviewer/viewer.fcgi?id=NC_045512.2&db=nuccore&report=fasta&retmode=text&withmarkup=on&tool=portal&log$=seqview&maxdownloadsize=100000000) and the respective [annotation `gff`](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/858/895/GCF_009858895.2_ASM985889v3/GCF_009858895.2_ASM985889v3_genomic.gff.gz). When downloading the dataset, an additional, smaller `fastq` file was included without the usual "_1" or "_2" suffix. For the purposes of this report, these have been removed along with all reverse reads.

## Sequence Data Quality
I ran FastQC on each downloaded `fastq` file to get a summary of the quality of each sample [@andrew2010fastqc]. Overall, all the samples passed `fastqc` per base sequence quality and per sequence quality checks.

&nbsp;
  
![An example of one sample's per base sequence quality FastQC report](/home/Gao_Sarah/sarah-gao-bioinformatics-final-project/data/images/SRR12768739_1_per_base_seq_qual.png)
*An example of one sample's FastQC quality report showing per base sequence quality.*
  
&nbsp;

![An example of one sample's FastQC per sequence quality scores report](/home/Gao_Sarah/sarah-gao-bioinformatics-final-project/data/images/SRR12768739_1_per_seq_qual.png)
*The FastQC per sequence quality scores of the same sample.*

&nbsp;
  
All samples had either a sequence length of 150bp or 76bp. Given the uniform distribution of the FastQC sequence length plot, one would assume that some processing had already occurred upstream to trim sequences of other lengths. It's possible, for example, that the sequencing machine itself threw out fragments during processing so as to keep only full-length reads.
  
&nbsp;

![An example of one sample's FastQC sequence length distribution](/home/Gao_Sarah/sarah-gao-bioinformatics-final-project/data/images/SRR12768739_1_seq_length_distr.png)
*The sequence length distribution report of the same sample.*

&nbsp;
  
## Data Processing

I used Trimmomatic on each downloaded `fastq` file to throw out bad sequences and trim areas of poor quality [@bolger2014trim]. For faster turnaround, I used TrimmomaticSE for single ends rather then using a paired end approach. For my parameters, I used 4 computing threads, converted quality scores to Phred-33, had a leading and trailing value of 5 bases, and a sliding window of size 8 with a minimum phred score of 25. Reads below a length of 100bp were dropped. Of the 102 sequences, 73 did not clear the Trimmomatic steps and I proceeded with 29 sequences.

I then aligned each read against the reference genome using the Burrows-Wheeler Aligner [@li2013bwa]. These aligned `bam` files were then converted to `sam` and then sorted using samtools [@li2009samtools]. I ran `bcftools mpileup` to determine read coverage of positions in the genome per base and then `bcftools call` to call the single nucleotide polymorphisms (SNPs) for each file. Finally, I filtered each file to remove short variants and retain only major variants. 
  
\newpage
  
# Results and Discussion

## Beijing Sequence Data

### Number of Mutations

By mapping the variants to the gene locations as identified by the Wuhan reference genome, I found that there were more SNPs in the samples in the N and S gene areas (Figure 2). However, if we examine the lengths of these genes, we find that the N and S genes are the longest (Table 1). To understand if there were indeed more mutations in these regions, I calculated the expected number of variants within each gene assuming equal rates of mutation based on their length proportional to the total length of all identified genes (Figure 2). From this, we can see that observed SNPs within the N gene is indeed higher than what we would expect with equal mutation rates. The S gene, on the other hand, actually had fewer variants than expected. Significance was not been calculated for these values.

### Mapping Mutations

Previous studies have identified positions at which mutations have been commonly found in samples taken around the world. Some have been found in specific geographic regions, suggesting regional strains with distinctive mutation patterns [@pachetti2020].

One strain in particular has emerged in prominence around the world, usurping the previously dominant D614 strain. This increased prevalence arose at multiple geographic levels globally, suggesting that this variant confers an evolutionary advantage over the previously established strain and has been positively selected for. This variant has a change within the Spike protein encoding gene, changing amino acid D614 to G614 through a A-to-G mutation at position 23403. G614 variants almost always also have 3 other distinguishing mutations within coding regions: C14408T within the ORF1b gene that causes an amino acid change in RNA-dependent RNA polymerase (RdRp P323L), C3037T within ORF1a which is silent, and C241T in the 5' UTR [@korber2020][@yang2020][@vanDorp2020]. Table 2 shows these common mutations and which genes they're found in.

I mapped the Beijing dataset's variants to these mutation positions and found that there were mutations characteristic of the D614G strain (Figure 3). Interestingly, the mutations characteristic of the D614G strain (241, 3037, 14408, and 23403) appeared with nearly the exact same frequency, with the exception being the mutation at 23403 having 6 occurrences rather than 5. This lends credence to the idea that these mutations occur in tandem and may serve as markers of the strain. That these have been found in samples from China also suggests that this variant may have become more prevalent despite the original strain of the Wuhan reference genome having already been previously established in the region. This is further supported by the work by Korber et al. examining the increased frequency of the D614G variant in Mainland China over time [@korber2020].

## Epidemiology

### Transmission Rates

While Korber et al. hypothesized that the increased frequency of the G614 strain was due to its higher rate of transmission, there is still no clear evidence that this variant is any more infectious or transmissible than the D614 strain [@grubaugh2020][@vanDorp2020]. It is evident, however, that transmission rates are affected by the behavior of the regional governments and populace.

China was the first country to report cases of SARS-CoV-2 infections and had spikes in cases much earlier on before the virus spread to other countries. In Figure 4, we can see how Beijing experienced a large spike in confirmed cases in January and February, a smaller one in March, and then another large one in June before plateauing [@mahdi2020oxcovid]. The large spike in January-February followed a national trend based on daily new case rates in China (Figure 5) [@owid2020]. China as a whole experienced a smaller spike in July as well.

Interestingly, when we examine the effective reproduction rate (R) of the virus throughout the year, the spikes in R seem to correlate with the spikes in case numbers we found in Beijing and across the whole country (Figure 6). Since the last spike in July, the R value has remained mostly under 1, correlating with the gradual decrease in case numbers seen in August.

### Government Response

The Chinese government's response to the pandemic may help to partially explain China's current low case numbers. The government put relatively severe restrictions on movement from late January to early February. Other researchers have shown that within-city movement and R were highly correlated during this time, suggesting a high efficacy of this response [@ainslie2020]. A strong and prolonged government response throughout the summer may also partially account for the drop in new cases as well as the prolonged period with R < 1. This lends support to the idea that transmission of the virus throughout a region can be mitigated by a stronger government response and adherence to social distancing practices.
  
\newpage
  
```{r load-packages-and-functions}
library("vcfR")
library("ggplot2")
library("dplyr")
library("ggthemes")
library("oxcovid19")
library("tidyr")
library("scales")
library("readr")
library("stringr")

# this loads all function files in the code/functions directory
function_paths <- list.files("code/functions", full.names = TRUE)
purrr::walk(function_paths, source)
```

```{r load-vcf-data}
# load in, tidy, and stack all of the VCF files in the target directory
# these are the output of the bash pipeline
stacked_vcfs <- parse_tidy_and_stack_vcfs(
  vcf_dir_path = params$vcf_dir_path)

# load in the gff file with genome annotations for SARS-CoV-2
gff <- read_gff(gff_file_path = params$gff_file_path)

# pull out just the named genes from the annotation file
gene_table <- extract_genes_from_gff(annotation_object = gff)

# combine the stacked vcf table with the gene names and the metadata from
# the sra runtable file
vcf_with_metadata <- add_genes_metadata_to_vcfstack(
  sra_runtable_path = params$sra_runtable_path,
  stacked_vcf = stacked_vcfs,
  cleaned_genes_table = gene_table)
```

# Tables

```{r gene-lengths-table}
# Creates a table to show the length of each gene using its start and end values
gene_table_lengths <- gene_table %>%
  mutate(length = end - start) %>%
  select(gene_name, start, end, length)
  
gene_table_lengths %>%
  knitr::kable(col.names = c("Gene Name",
                             "Start",
                             "End",
                             "Length"))
```

**Table 1**: Gene names, locations, and lengths in the SARS-CoV-2 genome.

&nbsp;
&nbsp;
&nbsp;
&nbsp;
  
```{r mutations-by-gene}
# Creates a table to show common mutations
common_mutations <- data.frame("mut_position" = c(
  1397, 2891, 3036, 8782, 11083, 17746, 17747, 17857, 18060,
  26143, 26144, 28144, 28881), "gene_name" = (NA), "type" = "other")
d614g_mutations <- data.frame("mut_position" = c(
  241, 3037, 14408, 23403), "gene_name" = (NA), "type" = "d614g")

key_mutations <- rbind(common_mutations, d614g_mutations) %>%
  arrange(mut_position)

# Match key mutations vector with genes table
for (n in 1:dim(key_mutations)) {
  for (row in 1:dim(gene_table)) {
    if (key_mutations$mut_position[n] >= gene_table$start[row] &
        key_mutations$mut_position[n] <= gene_table$end[row]) {
          key_mutations$gene_name[n] <- gene_table$gene_name[row]
    }
  }
}

# Append known gene names from articles
key_mutations$gene_name[1] <- "5' UTR"
key_mutations$gene_name[2:7] <- "ORF1a"
key_mutations$gene_name[8:12] <- "ORF1b"

# Make table
key_mutations %>%
  select(mut_position, gene_name) %>%
  knitr::kable(col.names = c("Mutation Position",
                             "Gene Name"))

```

**Table 2**: Mutation positions and the genes they are found within [@pachetti2020][@korber2020][@yang2020]. ***Note:*** Positions 3036, 17746, and 26143, which were identified by Pachetti et al., were not identified in the analyses by van Dorp et al. Instead, adjacent position 3037, 17747, and 26144 were identified, with a relatively large number of isolates having the mutation with high quality reads, which have been included here.
  
\newpage
  
# Figures

&nbsp;
&nbsp;
&nbsp;
&nbsp;
  
```{r set-plot-themes}
theme_update(plot.title = element_text(face = "bold",
                                       size = 12,
                                       hjust = 0.5,
                                       margin = margin(10, 0, 10, 0),
                                       lineheight = 1.2),
             axis.title.x = element_text(size = 10,
                                         face = "bold",
                                         vjust = -3),
             axis.text.x = element_text(size = 8,
                                        color = "#808080"),
             axis.title.y = element_text(size = 10,
                                         face = "bold",
                                         vjust = 3),
             axis.text.y = element_text(size = 8,
                                        color = "#808080"),
             plot.margin = margin(20, 30, 40, 30),
             legend.position = "bottom",
             legend.title = element_text(size = 6, 
                                         face = "bold", 
                                         color = "#808080",
                                         margin = margin(r = 10, unit = "pt")),
             legend.text = element_text(size = 6, 
                                        color = "#808080",
                                        margin = margin(r = 10, unit = "pt")),
             legend.background = element_rect(color = "#E7E7E7"),
             legend.margin = margin(5, 5, 5, 5),
             legend.box.spacing = unit(20, "pt"),
             legend.key.size = unit(8, "pt")
             )
```

```{r samples-by-isolation-source}
# Create a bar graph showing the distribution of sample isolation sources
isolation_source <- vcf_with_metadata %>%
  group_by(Isolation_source) %>%
  tally()

ggplot(isolation_source, aes(x = Isolation_source,
                              y = n)) +
  geom_col(fill = "#F8766D") +
  labs(title = "Number of Samples per Isolation Source",
       x = "Isolation Source",
       y = "Number of Samples") +
  scale_x_discrete(labels = c("Feces", "Pharyngeal Swab", "Sputum"))

# Save figure to the output figures directory as a .png
# ggsave(plot = isolation_source_plot,
#       filename = "output/figures/isolation_source_plot.png")
```
  
**Figure 1**: Samples were isolated from, in order from most to fewest, sputum, fecal, and pharyngeal swab sources.

&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
  
```{r expected-v-observed-variants}
# Create a plot that compares the number of SNPs found in each gene with the
# expected number of SNPs if equal rates of mutation across all genes

# Creates dataframe to store unique SNP counts
unique_snps <- vcf_with_metadata %>%
  filter(!is.na(gene)) %>%
  group_by(gene, pos) %>%
  tally() %>%
  group_by(gene) %>%
  tally()

# Create vector with percentages of gene length to total genome length
total_genes_length <- sum(gene_table_lengths$length)
unique_snps_per <- matrix(NA, 0, 3)
gene_per <- matrix(NA, 0, 3)
gene_per <- gene_table_lengths %>%
  mutate(length = end - start) %>%
  mutate(percentage = length / total_genes_length) %>%
  select(gene_name, length, percentage)

# Reorder percentages dataframe by gene name to match
gene_per <- gene_per[order(gene_per$gene_name),]

# Add a column in unique SNPs dataframe that shows percentage
total_snps = sum(unique_snps$n)
unique_snps_per <- cbind(unique_snps,
                         'percentage_of_total_genes' = gene_per$percentage,
                         'expected_snps'= total_snps*gene_per$percentage)
unique_snps_per <- unique_snps_per %>%
  pivot_longer(
  cols = c('n', 'expected_snps'),
  names_to = "expect_v_actual",
  values_to = "snp_values")

# Create a grouped bar graph comparing expected vs actual SNP counts
ggplot(unique_snps_per, aes(fill = expect_v_actual,
                            x = gene,
                            y = snp_values)) +
      geom_col(position = "dodge") +
      labs(title = paste("Count of Distinct SNPs in \n",
           "Named SARS-CoV-2 Genes"),
          x = "Gene Name",
          y = "SNP Values") +
      theme(legend.title = element_blank()) +
      scale_fill_discrete(labels = c("Expected Counts", "Observed Counts"))

# Save figure to the output figures directory as a .png
# ggsave(plot = exp_v_obs_plot,
#       filename = "output/figures/exp_v_obs_plot.png")
  
```
  
**Figure 2**: Comparison of expected SNP counts and observed SNP counts based on equal rate of mutation across all genes.

&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
  
```{r number-of-common-mutations}
# Create a bar graph that shows the number of samples of this dataset with
# mutations at the positions identified by previous studies
key_mutations_tallied <- stacked_vcfs %>%
  subset(pos %in% key_mutations$mut_position) %>%
  group_by(pos) %>%
  tally()

key_mutations_types <- key_mutations %>%
  subset(mut_position %in% key_mutations_tallied$pos) %>%
  group_by(mut_position) %>%
  select(type)

key_mutations_all <- merge(
  key_mutations_tallied, key_mutations_types,
  by.x = "pos",
  by.y = "mut_position")

key_mutations_all$pos <- factor(
  key_mutations_all$pos, levels = key_mutations_all$pos)

key_mutations_all %>%
  ggplot(aes(x = pos, y = n)) +
    geom_col(aes(fill = type)) +
    labs(title = "Number of Samples with Common Mutations",
         x = "Position Number",
         y = "Number of Sequences",
         fill = "Associated Strain") +
    scale_fill_discrete(labels = c("D614G", "Other")) +
    scale_y_continuous(breaks = pretty_breaks())

# Save figure to the output figures directory as a .png
# ggsave(plot = key_mutations_plot,
#       filename = "output/figures/key_mutations_plot.png")
```
  
**Figure 3**: Number of samples from this specific Beijing data set with common mutations found in samples globally. ***Note:*** Positions 3036, 17746, and 26143, which were identified by Pachetti et al., were not identified in the analyses by van Dorp et al [@pachetti2020][@korber2020]. Instead, adjacent position 3037, 17747, and 26144 were identified, with a relatively large number of isolates having the mutation with high quality reads, which have been included here.

&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
  
```{r beijing-confirmed-cases}
# Create a plot showing the total number of confirmed cases over time

# Create a connection to the Oxford COVID-19 database
con <- connect_oxcovid19()

# Access China's epidemiology table
epi_tab <- get_table(con = con, tbl_name = "epidemiology")
chn_epi_tab <- dplyr::filter(.data = epi_tab, countrycode == "CHN")

# Filter table by Beijing only
beijing_epi_tab <- chn_epi_tab %>%
  filter(adm_area_1 == "Beijing")

# Create plot of confirmed cases
beijing_epi_tab %>%
  select(date, adm_area_1, confirmed) %>%
  collect(beijing_epi_tab) %>%
  ggplot(aes(x = date,
             y = confirmed)) +
  labs(x = "Date",
       y = "Number of Confirmed Cases",
       title = paste("Total Number of Confirmed Cases of \n",
       "COVID-19 in Beijing Over Time")) +
  geom_line(color = "#00BFC4") +
  scale_x_date(breaks = waiver(), date_labels = "%b", date_breaks = "1 month")

# Save figure to the output figures directory as a .png
# ggsave(plot = beijing_total_confirmed_plot,
#       filename = "output/figures/beijing_total_confirmed_plot.png")
```
  
**Figure 4**: Total confirmed cases in Beijing over time, starting from January 20th, 2020.

&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
  
```{r china-new-cases-daily}
# Create a plot showing the number of daily new cases in China using Our World
# in Data dataset

# Get data from Our World in Data
ourworld_epi <- read_csv(
  "https://covid.ourworldindata.org/data/owid-covid-data.csv")

chn_ourworld_tab <- ourworld_epi %>%
  filter(location == "China")

chn_new_cases <- chn_ourworld_tab %>%
  select(date, new_cases, new_cases_smoothed) %>%
  pivot_longer(
    cols = c("new_cases", "new_cases_smoothed"), 
    names_to = "raw_v_smooth")

ggplot(chn_new_cases, 
       aes(x = date, y = value, color = raw_v_smooth)) +
  geom_line() +
  labs(x = "Date",
       y = "Number of New Cases",
       title = "Daily New Cases of COVID-19 in China per Day") +
  scale_color_discrete(labels = str_wrap(c("Daily New Cases", 
                                "Daily New Cases (7-Day Smoothed)"), 
                                width = 20)) +
  scale_x_date(breaks = waiver(), date_labels = "%b", date_breaks = "1 month") +
  theme(legend.title=element_blank())

# Save figure to the output figures directory as a .png
# ggsave(plot = china_daily_cases_plot,
#       filename = "output/figures/china_daily_cases_plot.png")
```
  
**Figure 5**: New daily confirmed cases over time in China. The red line represents the raw number of daily new cases whereas the blue line represents daily confirmed cases with 7-day smoothing.

&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
  
``` {r reproduction-rate-over-time}
# Create a plot that shows the effective reproduction rate over time in China
chn_ourworld_tab %>%
  ggplot(aes(x = date,
             y = reproduction_rate)) +
  geom_line(color = "#F8766D") +
  labs(x = "Date",
       y = "Effective Reproduction Rate (R)",
       title = "Effective Reproduction Rate (R) of SARS-CoV-2 in China") +
  scale_x_date(breaks = waiver(), date_labels = "%b", date_breaks = "1 month")

# Save figure to the output figures directory as a .png
# ggsave(plot = china_r_rate,
#       filename = "output/figures/china_r_rate.png")
```
  
**Figure 6**: Effective reproduction rate of the SARS-CoV-2 virus in China over time. R indicates how many new infections one infected person causes on average. If R is below 1, the number of cases will gradually decrease whereas if it's greater than 1, cases will increase.

&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
&nbsp;
  
``` {r stringency-index-over-time}
# Create a plot showing the stringency index, a measure of how stringent a
# government's response is to the pandemic, over time in China
chn_ourworld_tab %>%
  ggplot(aes(x = date,
             y = stringency_index)) +
  geom_line(color = "#00BFC4") +
  labs(x = "Date",
       y = "Stringency Index",
       title = "The Chinese Government's Response to the COVID-19 Pandemic") +
  scale_x_date(breaks = waiver(), date_labels = "%b", date_breaks = "1 month") +
  scale_y_continuous(limits = c(NA,100))

# Save figure to the output figures directory as a .png
# ggsave(plot = china_stringency_plot,
#       filename = "output/figures/china_stringency_plot.png")
```
  
**Figure 7**: The Chinese government's response to the pandemic measured by a stringency index over time. This index is compositely based on 9 response indicators, including school and workplace closures, contact tracing, and face coverings, and is scaled from 0 to 100 with 100 being the strictest response [@owid2020].
  
\newpage
  
# Sources Cited
