---
title: "Tracking SARS-CoV-2 mutation in Beijing 11 months after Wuhan onset"
author: "Sarah Gao"
date: "November 19, 2020"
output: pdf_document
csl: bioinformatics.csl
bibliography: references.bib
params:
  gff_file_path: "/data/sars_vcf_analysis/02_genome_reference/sars_refgenome_annotation.gff"
  vcf_dir_path: "data/11_vcf_output_for_R"
  sra_runtable_path: "data/00_sra_runtable/SraRunTable_PRJNA667180.txt"
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      warning = FALSE,
                      message = FALSE,
                      error = TRUE)
```

# Background and Overview

This is a report examining how the SARS-CoV-2 virus has mutated within one country over time and includes some variant analysis [@koyama2020variant].
This report uses a data set taken from sampling patients in Beijing with positive COVID-19 diagnoses from Beijing Ditan Hospital Capital Medical University. These samples were taken on October 3rd and 4th of 2020 and are compared against the Wuhan reference genome that was collected in December of 2019.

This report then examines how case rates and transmission rates in Beijing have fluctuated over time as the virus has continued to spread throughout the world.

# Methods

Note: When downloading the data set, an additional, smaller fastq file is included without the usual "_1" or "_2" suffix. These have been removed along with the reverse reads.



See the set of tutorials on the [vcfR package website](https://knausb.github.io/vcfR_documentation/index.html).

You may also want to use any of a range of different COVID data packages and data sources:

* https://kjhealy.github.io/covdata/
* https://github.com/como-ph/oxcovid19
* https://ropensci.org/blog/2020/10/20/searching-medrxivr-and-biorxiv-preprint-data/
* https://covidtracking.com/data/api
  * `readr::read_csv("https://api.covidtracking.com/v1/states/daily.csv")`
* https://rt.live/
  * `readr::read_csv("https://d14wlfuexuxgcm.cloudfront.net/covid/rt.csv")`

## Subsections are ok too

# Results and Discussion

```{r load-packages-and-functions}
library("vcfR")
library("ggplot2")
library("dplyr")
library("ggthemes")
library("oxcovid19")

# this loads all function files in the code/functions directory
function_paths <- list.files("code/functions", full.names = TRUE)
purrr::walk(function_paths, source)
```

```{r load-vcf-data}
# load in, tidy, and stack all of the VCF files in the target directory
# these are the output of the bash pipeline
stacked_vcfs <- parse_tidy_and_stack_vcfs(
  vcf_dir_path = params$vcf_dir_path)

# load in the gff file with genome annotations for SARS-CoV-2
gff <- read_gff(gff_file_path = params$gff_file_path)

# pull out just the named genes from the annotation file
gene_table <- extract_genes_from_gff(annotation_object = gff)

# combine the stacked vcf table with the gene names and the metadata from
# the sra runtable file
vcf_with_metadata <- add_genes_metadata_to_vcfstack(
  sra_runtable_path = params$sra_runtable_path,
  stacked_vcf = stacked_vcfs,
  cleaned_genes_table = gene_table)
```

# Tables

```{r example-table}
# An example table to show the length of each gene using its start and end
gene_table_lengths <- gene_table %>%
  mutate(length = end - start) %>%
  select(gene_name, start, end, length)

gene_table_lengths %>%
  knitr::kable(col.names = c("Gene Name",
                             "Start",
                             "End",
                             "Length"))
```

**Table 1**: Gene names, locations, and lengths in the SARS-CoV-2 genome. Higher SNP counts in the S and N genes may be related to the larger size of these genes.

# Figures

```{r samples-by-isolation-source}
# Create a bar graph showing the distribution of sample isolation sources
isolation_source <- vcf_with_metadata %>%
  group_by(Isolation_source) %>%
  tally()

ggplot(isolation_source, aes(x = Isolation_source,
                              y = n)) +
  geom_col() +
  labs(title = "Number of Samples per Isolation Source",
       x = "Isolation Source",
       y = "Number of Samples")
```

```{r SNPS-by-genes}
# Create a plot of unique SNP locations within each gene across all samples
unique_snps <- vcf_with_metadata %>%
  filter(!is.na(gene)) %>% # get rid of SNPs not in gene regions
  group_by(gene, pos) %>%
  tally() %>% # this gives a column n for the number of gene by position
  group_by(gene) %>%
  tally() # this collapses that down to the number of unique SNP locations

ggplot(unique_snps, aes(x = gene,
                        y = n)) +
      geom_col() +
      labs(title = "Count of distinct SNPs in Named SARS-CoV-2 Genes",
          x = "Gene Name") +
      theme_few() # get rid of the grey background
```

**Figure 1**: N and S genes have more unique SNPs in the set of samples analyzed.

```{r variants-by-percentage}
# Create a plot that compares the number of SNPs found in each gene with the
# expected number of SNPs if proportional to gene length
# Create vector with percentages of gene length to total genome length
total_genes_length <- sum(gene_table_lengths$length)
unique_snps_per <- matrix(NA, 0, 3)
gene_per <- matrix(NA, 0, 3)
gene_per <- gene_table_lengths %>%
  mutate(length = end - start) %>%
  mutate(percentage = length / total_genes_length) %>%
  select(gene_name, length, percentage)

# Reorder percentages dataframe by gene name to match
gene_per <- gene_per[order(gene_per$gene_name),]

# Add a column in unique SNPs dataframe that shows percentage
total_SNPs = sum(unique_snps_per$n)
unique_snps_per <- cbind(unique_snps,
                         'percentage_of_total_genes' = gene_per$percentage, 'expected_SNPs' = total_SNPs*unique_snps_per$percentage_of_total_genes)


# Create plot
ggplot(unique_snps_per, aes(x = gene,
                        y = weighted_SNPs)) +
      geom_col() +
      labs(title = "Count of distinct SNPs in Named SARS-CoV-2 Genes",
          x = "Gene Name") +
      theme_few() # get rid of the grey background)
  
```
# Sources Cited
