---
title: "Identifying SARS-CoV-2 mutations in Beijing and transmission rates in China in 2020"
author: "Sarah Gao"
date: "November 19, 2020"
output:
  pdf_document: default
  html_document:
    df_print: paged
csl: bioinformatics.csl
bibliography: references.bib
params:
  gff_file_path: /data/sars_vcf_analysis/02_genome_reference/sars_refgenome_annotation.gff
  vcf_dir_path: data/11_vcf_output_for_R
  sra_runtable_path: data/00_sra_runtable/SraRunTable_PRJNA667180.txt
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      warning = FALSE,
                      message = FALSE,
                      error = TRUE)
```

# Background and Overview

In late 2019, the first cases of Severe Acute Respiratory Syndrome Coronavirus 2 caused by the SARS-CoV-2 virus were reported in Wuhan, China. Since then, the virus has spread rapidly around the world and has resulted in more than 1.4 million deaths worldwide as of November, 27th 2020 [@jhu2020].

Researchers around the globe have identified key mutations that are predominant in different locations, indicating that the virus has been evolving over time to form location-specific strains. This report examined how the SARS-CoV-2 virus has mutated within and affected the population in Beijing, China. Beijing is approximately 717 miles (~1,153 km) from Wuhan, a city in Hubei province that was the center of the first known COVID-19 cases first reported in December 2019.

The goal of this report was to identify mutations of the SARS-CoV-2 virus found in samples taken from positive patients in a Beijing hospital, compare them with known mutations identified by previous researchers, and analyze data on China's response to the virus at a national level. I used a pipeline written in bash to process the sequence data and identify variants from the reference Wuhan genome. I called out specifically variants at positions that are commonly found in samples around the world in order to examine how Beijing's strain may have mutated since the virus first emerged [@pachetti2020][@vanDorp2020].

Previous studies showed that a mutation in the gene encoding the S spike protein has created a SARS-CoV-2 variant that has become the most prevalent form in many geographies, suggesting that this strain could have an evolutionary advantage [@korber2020]. While there has been no definitive understanding in how known mutations in the virus affect the virus' transmission or infection capacity, it is certainly an important area of study to understand how the virus will change over time, and in turn, how we may more effectively combat it. Analyses of this Beijing dataset show that there were no more S gene mutations than one would expect assuming equal rates of mutation across the genome. However, there are many other factors that can affect how a region copes with the ever-changing COVID-19 pandemic other than the potential outcomes of viral mutations, including population and government responses. I found that the number of cases across the country as a whole has reduced dramatically since spring of 2020, despite fluctuating effective reproduction rates, as the Chinese government has reacted quickly and strictly in response to spikes. 


# Methods

This report used a data set taken from sampling patients in Beijing with positive COVID-19 diagnoses and includes some variant analysis [@koyama2020variant]. Du et al took these samples from SARS-CoV-2 positive patients at Beijing Ditan Hospital Capital Medical University [@du2020] via three isolation sources: feces, pharyngeal swab, and sputum (Figure 1). The 

Note: When downloading the data set, an additional, smaller fastq file is included without the usual "_1" or "_2" suffix. These have been removed along with the reverse reads.

See the set of tutorials on the [vcfR package website](https://knausb.github.io/vcfR_documentation/index.html).

You may also want to use any of a range of different COVID data packages and data sources:

* https://kjhealy.github.io/covdata/
* https://github.com/como-ph/oxcovid19
* https://ropensci.org/blog/2020/10/20/searching-medrxivr-and-biorxiv-preprint-data/
* https://covidtracking.com/data/api
  * `readr::read_csv("https://api.covidtracking.com/v1/states/daily.csv")`
* https://rt.live/
  * `readr::read_csv("https://d14wlfuexuxgcm.cloudfront.net/covid/rt.csv")`

## Subsections are ok too

# Results and Discussion

This variant has a change within the Spike protein encoding gene, changing amino acid D614G to G614. This increased prevalence arose at multiple geographic levels globally, suggesting that the G614 variant confers an evolutionary advantage over the previously established strain and has been positively selected for. 

```{r load-packages-and-functions}
library("vcfR")
library("ggplot2")
library("dplyr")
library("ggthemes")
library("oxcovid19")
library("tidyr")
library("scales")
library("readr")
library("stringr")

# this loads all function files in the code/functions directory
function_paths <- list.files("code/functions", full.names = TRUE)
purrr::walk(function_paths, source)
```

```{r load-vcf-data}
# load in, tidy, and stack all of the VCF files in the target directory
# these are the output of the bash pipeline
stacked_vcfs <- parse_tidy_and_stack_vcfs(
  vcf_dir_path = params$vcf_dir_path)

# load in the gff file with genome annotations for SARS-CoV-2
gff <- read_gff(gff_file_path = params$gff_file_path)

# pull out just the named genes from the annotation file
gene_table <- extract_genes_from_gff(annotation_object = gff)

# combine the stacked vcf table with the gene names and the metadata from
# the sra runtable file
vcf_with_metadata <- add_genes_metadata_to_vcfstack(
  sra_runtable_path = params$sra_runtable_path,
  stacked_vcf = stacked_vcfs,
  cleaned_genes_table = gene_table)
```

# Tables

```{r gene-lengths-table}
# An example table to show the length of each gene using its start and end
gene_table_lengths <- gene_table %>%
  mutate(length = end - start) %>%
  select(gene_name, start, end, length)
  
gene_table_lengths %>%
  knitr::kable(col.names = c("Gene Name",
                             "Start",
                             "End",
                             "Length"))
```

**Table 1**: Gene names, locations, and lengths in the SARS-CoV-2 genome. Higher SNP counts in the S and N genes may be related to the larger size of these genes.

```{r mutations-by-gene}
# Create a table to show the mutation positions and associated genes identified
# by Pachetti et al. (2020)
key_mutations <- data.frame("mut_position" = c(
  1397, 2891, 3036, 3037, 8782, 11083, 14408, 17746, 17747, 17857, 18060, 23403, 
  26143, 26144, 28144, 28881), "gene_name" = (NA))

# Match key mutations vector with genes table
for (n in 1:dim(key_mutations)) {
  for (row in 1:dim(gene_table)) {
    if (key_mutations$mut_position[n] >= gene_table$start[row] &
        key_mutations$mut_position[n] <= gene_table$end[row]) {
          key_mutations$gene_name[n] <- gene_table$gene_name[row]
    }
  }
}

# Create table
key_mutations %>%
  knitr::kable(col.names = c("Mutation Position",
                             "Gene Name"))

```
**Table 2**: Mutation positions and the genes they are found within as identified by Pachetti et al in all geographic locations [@pachetti2020emerging]. ***Note:*** Positions 3036, 17746, and 26143, which were identified by Pachetti et al., were not identified in the analyses by van Dorp et al. Instead, adjacent position 3037, 17747, and 26144 were identified, with a relatively large number of isolates having the mutation with high quality reads, which have been included here.

# Figures

```{r samples-by-isolation-source}
# Create a bar graph showing the distribution of sample isolation sources
isolation_source <- vcf_with_metadata %>%
  group_by(Isolation_source) %>%
  tally()

ggplot(isolation_source, aes(x = Isolation_source,
                              y = n)) +
  geom_col() +
  labs(title = "Number of Samples per Isolation Source",
       x = "Isolation Source",
       y = "Number of Samples") +
  scale_x_discrete(labels = c("Feces", "Pharyngeal Swab", "Sputum")) +
  theme_few()
```
**Figure 1**: Samples were isolated from, in order from most to fewest, sputum, fecal, and pharyngeal swab sources.

```{r quality-scores}
# Create a plot showing the quality scores of the reads


```

Here we see the SNP counts in each gene region and that indeed, there are more SNPs in the larger S and N genes.
```{r SNPS-by-genes}
# Create a plot of unique SNP locations within each gene across all samples
unique_snps <- vcf_with_metadata %>%
  filter(!is.na(gene)) %>% # get rid of SNPs not in gene regions
  group_by(gene, pos) %>%
  tally() %>% # this gives a column n for the number of gene by position
  group_by(gene) %>%
  tally() %>% # this collapses that down to the number of unique SNP locations
  rename(observed_snps = n)

ggplot(unique_snps, aes(x = gene,
                        y = observed_snps)) +
      geom_col() +
      labs(title = "Count of distinct SNPs in Named SARS-CoV-2 Genes",
          x = "Gene Name",
          y = "SNP Counts") +
      theme_few() # get rid of the grey background
```
**Figure 1**: N and S genes have more unique SNPs in the set of samples analyzed.


```{r variants-by-percentage}
# Create a plot that compares the number of SNPs found in each gene with the
# expected number of SNPs if proportional to gene length
# Create vector with percentages of gene length to total genome length
total_genes_length <- sum(gene_table_lengths$length)
unique_snps_per <- matrix(NA, 0, 3)
gene_per <- matrix(NA, 0, 3)
gene_per <- gene_table_lengths %>%
  mutate(length = end - start) %>%
  mutate(percentage = length / total_genes_length) %>%
  select(gene_name, length, percentage)

# Reorder percentages dataframe by gene name to match
gene_per <- gene_per[order(gene_per$gene_name),]

# Add a column in unique SNPs dataframe that shows percentage
total_snps = sum(unique_snps$observed_snps)
unique_snps_per <- cbind(unique_snps,
                         'percentage_of_total_genes' = gene_per$percentage,
                         'expected_snps'= total_snps*gene_per$percentage)
unique_snps_per <- unique_snps_per %>%
  pivot_longer(
  cols = c('observed_snps', 'expected_snps'),
  names_to = "expect_v_actual",
  values_to = "snp_values"
)

# Create a grouped bar graph comparing expected vs actual SNP counts
ggplot(unique_snps_per, aes(fill = expect_v_actual,
                            x = gene,
                            y = snp_values)) +
      geom_col(position = "dodge") +
      labs(title = "Count of distinct SNPs in Named SARS-CoV-2 Genes",
          x = "Gene Name",
          y = "SNP Values") +
  scale_fill_discrete(name = "Legend", labels = c("Expected counts",
                                                  "Observed counts")) +
      theme_few()
  
```
**Figure 2**: Comparison of expected SNP counts and observed SNP counts based on equal rate of mutation across all genes.


```{r number-of-common-mutations}
# Create a bar graph that shows the number of samples of this dataset with
# mutations at the positions identified by Pachetti et al.
key_mutations_tallied <- stacked_vcfs %>%
  subset(pos %in% key_mutations$mut_pos) %>%
  group_by(pos) %>%
  tally()

key_mutations_tallied$pos <- factor(key_mutations_tallied$pos, levels = key_mutations_tallied$pos)

key_mutations_tallied %>%
  ggplot(aes(x = pos, y = n)) +
    geom_col() +
    labs(title = "Number of samples with specific mutation",
         x = "Position number",
         y = "Number of samples") +
    scale_y_continuous(breaks = pretty_breaks())
```
**Figure 4**: Number of samples from this specific Beijing data set with common mutations found in samples globally. ***Note:*** Positions 3036, 17746, and 26143, which were identified by Pachetti et al., were not identified in the analyses by van Dorp et al. Instead, adjacent position 3037, 17747, and 26144 were identified, with a relatively large number of isolates having the mutation with high quality reads, which have been included here.

```{r beijing-confirmed-cases}
# Create a plot showing the number of confirmed cases over time

# Create a connection to the Oxford COVID-19 database
con <- connect_oxcovid19()

# Access China's epidemiology table
epi_tab <- get_table(con = con, tbl_name = "epidemiology")
chn_epi_tab <- dplyr::filter(.data = epi_tab, countrycode == "CHN")

# Filter table by Beijing only
beijing_epi_tab <- chn_epi_tab %>%
  filter(adm_area_1 == "Beijing")

# Retrieve remote table data into R
dplyr::collect(beijing_epi_tab)

# Create plot of confirmed cases
beijing_epi_tab %>%
  select(date, adm_area_1, confirmed) %>%
  ggplot(aes(x = date,
             y = confirmed)) +
  geom_line(color = "#00BFC4")
```
**Figure 5**: Total confirmed cases in Beijing over time, starting from January 20th, 2020.

```{r china-new-cases-daily}
# Create a plot showing the number of daily new cases in China
chn_new_cases <- chn_ourworld_tab %>%
  select(date, new_cases, new_cases_smoothed) %>%
  pivot_longer(
    cols = c("new_cases", "new_cases_smoothed"), 
    names_to = "raw_v_smooth")

ggplot(chn_new_cases, 
       aes(x = date, y = value, color = raw_v_smooth)) +
  geom_line() +
  scale_color_discrete(labels = str_wrap(c("Daily New Cases", 
                                "Daily New Cases (7-Day Smoothed)"), 
                                width = 20)) +
  theme(legend.title=element_blank())
```
**Figure 6**: New daily confirmed cases over time in China. The red line represents the raw number of daily new cases whereas the blue line represents daily confirmed cases with 7-day smoothing.

``` {r reproduction-rate-over-time}
# Create a plot that shows the effective reproduction rate over time in China
chn_ourworld_tab %>%
  ggplot(aes(x = date,
             y = reproduction_rate)) +
  geom_line(color = "#00BFC4")
```
**Figure 7**: Effective reproduction rate of the SARS-CoV-2 virus in China over time.

``` {r stringency-index-over-time}
# Create a plot showing the stringency index, a measure of how stringent a
# government's response is to the pandemic, over time in China
chn_ourworld_tab %>%
  ggplot(aes(x = date,
             y = stringency_index)) +
  geom_line(color = "#00BFC4")
```
**Figure 8**: The Chinese government's response to the pandemic measured by a stringency index over time. This index is based on 9 response indicators, including school and workplace closures, and is scaled from 0 to 100 with 100 being the strictest response [@owid2020].

# Sources Cited
